web_poet/overrides.py

```diff
@@ -37,6 +38,7 @@ class OverrideRule:
           about the patterns.
         * ``use`` - The Page Object that will be **used**.
         * ``instead_of`` - The Page Object that will be **replaced**.
+        * ``data_type`` - The data container class that the Page Object returns.
```

Comment:
I think that's fine to have different argument names in ApplyRule and handle_urls.

handle_urls parameters tell you something about the decorated page object:

```py
# MyPage handles "example.com" urls and overrides GenericProductPage:
@handle_urls("example.com", overrides=GenericProductPage)
class MyPage(ItemPage[Product]):
    # ...
```

ApplyRule attributes describe the rule:

```py
# for example.com website use MyPage instead of GenericProductPage
ApplyRule(for_patterns=Patterns(["example.com"]), use=Mypage, instead_of=GenericProductPage)
```

Following this logic, we could do something like this:

```py
# MyPage handles "example.com" urls and can be used to provide Product items:
@handle_urls("example.com", provides=Product)  # provides should rarely be needed
class MyPage(ItemPage[Product]):
    # ...


# for example.com website use MyPage to provide Product items
ApplyRule(for_patterns=Patterns(["example.com"]), use=MyPage, to_provide=Product)
```

In #77 it was `to_return` instead of `to_provide`; to_provide seems better.

---

After re-reading it though, the following example doesn't look too bad if we make argument names the same:

```py
# MyPage handles "example.com" instead of GenericProductPage:
@handle_urls("example.com", instead_of=GenericProductPage)
class MyPage(ItemPage[Product]):
    # ...

# MyPage handles "example.com" in order to provide Product items
@handle_urls("example.com", to_provide=Product)
class MyPage(ItemPage[Product]):
    # ...

```

Reply:
What do you think?

Reply:
+1 on having `@handle_urls()` utilize `instead_of`. This term is much more clear about the intent compared to the old `overrides`. The old `overrides` param speaks more about the a specific concept used in **web-poet**. The `instead_of` uses plain language which is direct. It also helps that the rule which `@handle_urls()` produce is consistent with its parameters.

I find that this is a good opportunity to improve the mismatches of parameters:

```python
# Old one
@handle_urls("example.com", overrides=GenericProductPage)
class MyPage(ItemPage[Product]):
    ...

ApplyRule(for_patterns=Patterns(["example.com"]), use=Mypage, instead_of=GenericProductPage)
```

From the old one above, there might be some confusion to newcomers regarding what happened to the class they passed to `overrides`. Is there an underlying logic which sets `overrides` apart from `instead_of`? Are they same?

This new one is more consistent which could help remove those types of confusions:

```python
# Proposed renaming
@handle_urls("example.com", instead_of=GenericProductPage)
class MyPage(ItemPage[Product]):
    ...

ApplyRule(for_patterns=Patterns(["example.com"]), use=Mypage, instead_of=GenericProductPage)
```

---

On the other hand, I find `to_return` to be slightly better than `to_provide`. It also helps that it's consistent in both `@handle_urls` and `ApplyRule`:

```python
@handle_urls("example.com", to_return=Product) 
class MyPage(ItemPage[Product]):
    ...

ApplyRule(for_patterns=Patterns(["example.com"]), use=MyPage, to_return=Product)
```

The `*_return` terminology is consistent with the behavior of the Page Object that it returns something. I think `*_provide` could also work. Although I think when it's used in the context of **scrapy-poet**, there might be some confusion that the _rule_ or _page object_ itself is the provider.

I might be missing something. What was the main benefits you see in using `to_provide` instead of `to_return`?

Reply:
I'm fine with using `to_return` name, it sounds fine indeed. `to_provide` may require more knowledge from the users, and so it could be less clear.

Reply:
Updated these. Thanks @kmike !